Bump SCRIPT_VERIFY flags to 64 bit

ajtowns opened the pull request:
We currently use 21 of 32 possible bits for `SCRIPT_VERIFY_*` flags, with open PRs that may use 8 more (#29247, #31989, #32247, #32453). The mutinynet fork that has included many experimental soft fork features is [already reusing bits here](https://github.com/benthecarman/bitcoin/blob/d4a86277ed8a0712e03fbbce290e9209165e049c/src/script/interpreter.h#L175-L195). Therefore, bump this to 64 bits.

In order to make it easier to update this logic in future, this PR also introduces a dedicated type for the script flags, and disables implicit conversion between that type and the underlying integer type. To make verifying that this change doesn't cause flags to disappear, this PR also resurrects the changes from #28806 so that the script flags that are consensus enforced on each block can be queried via getdeploymentinfo.

ajtowns committed:
Move mapFlagNames and FormatScriptFlags logic to script/interpreter.h

Moves FormatScriptFlags logic into GetScriptFlagNames which returns a
vector of strings. For completeness, also has GetScriptFlagNames report
on any bits that do not match a known script flag.

ajtowns committed:
rpc: have getdeploymentinfo report script verify flags

ajtowns committed:
validation: export GetBlockScriptFlags()

ajtowns committed:
script/interpreter: introduce script_verify_flags typename

Previously the SCRIPT_VERIFY_* flags were specified as either uint32_t,
unsigned int, or unsigned. This converts them to a common type alias in
preparation for changing the underlying type.

ajtowns committed:
script/interpreter: make script_verify_flag_name an ordinary enum

Instead of having `SCRIPT_VERIFY_FOO = (1U << n)` just have it
be `n` directly, and do the bit shifting when converting it to
`script_verify_flags`.

ajtowns committed:
script/verify_flags: extend script_verify_flags to 64 bits

ajtowns commented:
The "introduce script_verify_flags typename" commit is probably best reviewed with `--word-diff` fwiw.

cc @instagibbs @darosior @benthecarman

ajtowns commented:
CI failure seems to be due to [a bug in qt6 6.4](https://bugreports.qt.io/browse/QTBUG-31496?focusedId=888930&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-888930):

> Also for people coming here from Google: I had the same error message, but for me the problem was that I was using Qt 6.4.2 which apparently thinks that #if 202002L < 201103L is true, which causes c++config.h to not be included (and no #error is generated because moc doesn't support the directive) . This has been fixed somewhere before Qt 6.7.3.


```
[20:04:17.396] Get:287 http://archive.ubuntu.com/ubuntu noble/universe amd64 qt6-base-dev-tools amd64 6.4.2+dfsg-21.1build5 [970 kB]
[20:04:17.414] Get:288 http://archive.ubuntu.com/ubuntu noble/universe amd64 qt6-qpa-plugins amd64 6.4.2+dfsg-21.1build5 [89.3 kB]
[20:04:17.417] Get:289 http://archive.ubuntu.com/ubuntu noble/universe amd64 qt6-base-dev amd64 6.4.2+dfsg-21.1build5 [1458 kB]
...
[20:12:02.187] Command
[20:12:02.187] -------
[20:12:02.187] /usr/lib/qt6/libexec/moc -DBOOST_MULTI_INDEX_DISABLE_SERIALIZATION -DBOOST_NO_CXX98_FUNCTION_BASE -DQT_CORE_LIB -DQT_DBUS_LIB -DQT_GUI_LIB -DQT_NETWORK_LIB -DQT_NO_DEBUG -DQT_NO_KEYWORDS -DQT_USE_QSTRINGBUILDER -DQT_WIDGETS_LIB -I/ci_container_base/ci/scratch/build-x86_64-pc-linux-gnu/src -I/ci_container_base/src -I/ci_container_base/src/univalue/include -I/ci_container_base/src/leveldb/include -I/usr/include/x86_64-linux-gnu/qt6/QtWidgets -I/usr/include/x86_64-linux-gnu/qt6 -I/usr/include/x86_64-linux-gnu/qt6/QtCore -I/usr/lib/x86_64-linux-gnu/qt6/mkspecs/linux-g++ -I/usr/include/x86_64-linux-gnu/qt6/QtGui -I/usr/include/x86_64-linux-gnu/qt6/QtNetwork -I/usr/include/x86_64-linux-gnu/qt6/QtDBus -I/usr/include -I/usr/include/c++/13 -I/usr/include/x86_64-linux-gnu/c++/13 -I/usr/include/c++/13/backward -I/usr/lib/llvm-20/lib/clang/20/include -I/usr/local/include -I/usr/include/x86_64-linux-gnu --include /ci_container_base/ci/scratch/build-x86_64-pc-linux-gnu/src/qt/bitcoinqt_autogen/moc_predefs.h -p/ci_container_base/src/qt --output-dep-file -o /ci_container_base/ci/scratch/build-x86_64-pc-linux-gnu/src/qt/bitcoinqt_autogen/include/qt/intro.moc /ci_container_base/src/qt/intro.cpp
[20:12:02.187] 
[20:12:02.187] Output
[20:12:02.187] ------
[20:12:02.187] usr/include/c++/13/concept:46:1: error: Parse error at "std"
```

The error output also seems to be cutting off the last letter of the filename ("concept" vs "concepts" here, and "stl_relops." vs "stl_relops.h" in the original bug report linked above).

hebasto commented:
> CI failure seems to be due to [a bug in qt6 6.4](https://bugreports.qt.io/browse/QTBUG-31496?focusedId=888930&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-888930):

Although both failed CI job use Clang 20.1.7, the error can also be reproduced with GCC 13.3.

ajtowns commented:
> Minified this is:

I don't think you need the `git fetch; git checkout` here -- you get the same error with that command (dropping the `/bitcoin`) on master afaics?

ajtowns committed:
script/verify_flags: make script_verify_flags type safe

`using script_verify_flags = uint32_t` allows implicit conversion to
and from int, so replace it with a class to have the compiler ensure we
use the correct type. Provide from_int and as_int to allow for explicit
conversions when desired.

Introduces the type `script_verify_flag_name` for the individual flag
name enumeration.

hebasto commented:
> @hebasto how can we move past the GUI build failures here? https://github.com/bitcoin/bitcoin/actions/runs/16538448423/job/46776353144?pr=32998#step:6:3769.

Sure!

> > > CI failure seems to be due to [a bug in qt6 6.4](https://bugreports.qt.io/browse/QTBUG-31496?focusedId=888930&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-888930):
> > 
> > 
> > Although both failed CI job use Clang 20.1.7, the error can also be reproduced with GCC 13.3.
> 
> [Refactoring](https://github.com/hebasto/bitcoin/commits/pr32998/0717.refactored/) helps.

Implemented in https://github.com/bitcoin-core/gui/pull/881.

instagibbs reviewed (approved):
ACK 417437eb01ac014c57aca47f44d7f8d3da351987

non-blocking suggestions for coverage

ajtowns committed:
test: additional test coverage for script_verify_flags

darosior reviewed (approved):
ACK 652424ad162b63d73ecb6bd65bd26946e90c617f

I think it would also make sense, in a follow-up, to move the rest of the Script verification flags stuff to the new `verify_flags.h`.

theStack commented on src/script/verify_flags.h:
```diff
@@ -6,8 +6,66 @@
+    constexpr explicit operator bool() const { return m_value != 0; }
+    constexpr bool operator==(script_verify_flags other) const { return m_value == other.m_value; }
+
+    /** Compare two script_verify_flags. <, >, <=, and >= are auto-generated from this. */
+    friend constexpr std::strong_ordering operator<=>(const script_verify_flags& a, const script_verify_flags& b) noexcept
+    {
+        return a.m_value <=> b.m_value;
+    }
```
nit: `operator<` would be sufficient (currently only needed in the transaction tests, where a [`std::set` of script flag combinations](https://github.com/bitcoin/bitcoin/blob/86eaa4d6cd5c428f6635a8d44fa6b5a9545ea909/src/test/transaction_tests.cpp#L183) is used), I doubt that the other ones would ever have a use-case
```diff
diff --git a/src/script/verify_flags.h b/src/script/verify_flags.h
index 95a55d2c79..e14a329ace 100644
--- a/src/script/verify_flags.h
+++ b/src/script/verify_flags.h
@@ -47,11 +47,9 @@ public:
     // tests
     constexpr explicit operator bool() const { return m_value != 0; }
     constexpr bool operator==(script_verify_flags other) const { return m_value == other.m_value; }
-
-    /** Compare two script_verify_flags. <, >, <=, and >= are auto-generated from this. */
-    friend constexpr std::strong_ordering operator<=>(const script_verify_flags& a, const script_verify_flags& b) noexcept
+    friend constexpr bool operator<(const script_verify_flags& a, const script_verify_flags& b) noexcept
     {
-        return a.m_value <=> b.m_value;
+        return a.m_value < b.m_value;
     }

 private:
```

theStack reviewed (approved):
Code-review ACK 652424ad162b63d73ecb6bd65bd26946e90c617f :flags:
